### PR TITLE
Replace hostname and port with placeholders of callback URLs

### PR DESCRIPTION
**Pull request by Thisara-Welmilla**
Issue:
- https://github.com/wso2/product-is/issues/10514

Replace hostname and port with placeholders of callback URLs for only console and myaccount when storing in databases and resolve the callback URL at retrieval.

### CODE REVIEW

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java [hunk 1]
```diff
@@ -1026,4 +1031,40 @@ public static boolean isLegacyAuditLogsDisabledInAppMgt() {
...
+     *
+     * @param absoluteUrl     The absolute URL which need to be modified.
+     * @return The URL which origin replaced placeholders.
+     * @throws URLBuilderException If any error occurs when building absolute public url without path.
+     */
+    public static String replaceUrlOriginWithPlaceholders(String absoluteUrl) throws URLBuilderException {
+
+        if (StringUtils.contains(absoluteUrl, CONSOLE_APP_PATH) ||
+                StringUtils.contains(absoluteUrl, MY_ACCOUNT_APP_PATH)) {
+            String origin = ServiceURLBuilder.create().build().getAbsolutePublicUrlWithoutPath();
```
Comment (chamathns) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java:1046]: let's use `basePath`?

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java [hunk 2]
```diff
@@ -1026,4 +1031,40 @@ public static boolean isLegacyAuditLogsDisabledInAppMgt() {
...
+            absoluteUrl = StringUtils.replace(absoluteUrl, origin, BASE_URL_PLACEHOLDER);
+        }
+
```
Comment (chamathns) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java:1049]:
```suggestion
```

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java [hunk 3]
```diff
@@ -1026,4 +1031,40 @@ public static boolean isLegacyAuditLogsDisabledInAppMgt() {
...
+     * @throws URLBuilderException If any error occurs when building absolute public url without path.
+     */
+    public static String resolveOriginUrlFromPlaceholders(String absoluteUrl) throws URLBuilderException {
+
+        if (StringUtils.contains(absoluteUrl, CONSOLE_APP_PATH) ||
+                StringUtils.contains(absoluteUrl, MY_ACCOUNT_APP_PATH)) {
+            String origin = ServiceURLBuilder.create().build().getAbsolutePublicUrlWithoutPath();
+            absoluteUrl = StringUtils.replace(absoluteUrl, BASE_URL_PLACEHOLDER, origin);
+        }
+
```
Comment (chamathns) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java:1067]:
```suggestion
```

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java [hunk 4]
```diff
@@ -1026,4 +1031,40 @@ public static boolean isLegacyAuditLogsDisabledInAppMgt() {
...
+     * This method use to replace placeholders with the hostname and port of URLs of console and myaccount.
+     *
+     * @param absoluteUrl     The URL which need to resolve from placeholders.
+     * @return The resolved URL from placeholders.
+     * @throws URLBuilderException If any error occurs when building absolute public url without path.
+     */
+    public static String resolveOriginUrlFromPlaceholders(String absoluteUrl) throws URLBuilderException {
+
+        if (StringUtils.contains(absoluteUrl, CONSOLE_APP_PATH) ||
+                StringUtils.contains(absoluteUrl, MY_ACCOUNT_APP_PATH)) {
```
Comment (chamathns) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java:1063]: check from application name

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java [hunk 5]
```diff
@@ -1026,4 +1031,38 @@ public static boolean isLegacyAuditLogsDisabledInAppMgt() {
...
+     * This method use to replace placeholders with the hostname and port of URLs of console and myaccount.
+     *
+     * @param absoluteUrl     The URL which need to resolve from placeholders.
+     * @return The resolved URL from placeholders.
+     * @throws URLBuilderException If any error occurs when building absolute public url without path.
+     */
+    public static String resolveOriginUrlFromPlaceholders(String applicationName, String absoluteUrl) throws URLBuilderException {
+
+        if (applicationName.equalsIgnoreCase(CONSOLE_APP_NAME) ||
+                applicationName.equalsIgnoreCase(MY_ACCOUNT_APP_NAME)) {
```
Comment (chamathns) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java:1062]:
```suggestion
        if (CONSOLE_APP_NAME.equalsIgnoreCase(applicationName) ||
                MY_ACCOUNT_APP_NAME.equalsIgnoreCase(applicationName)) {
```

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java [hunk 6]
```diff
@@ -1026,4 +1031,38 @@ public static boolean isLegacyAuditLogsDisabledInAppMgt() {
...
+     * This method use to replace the hostname and port with placeholders of URLs of console and myaccount.
+     *
+     * @param absoluteUrl     The absolute URL which need to be modified.
+     * @return The URL which origin replaced placeholders.
+     * @throws URLBuilderException If any error occurs when building absolute public url without path.
+     */
+    public static String replaceUrlOriginWithPlaceholders(String applicationName, String absoluteUrl) throws URLBuilderException {
+
+        if (applicationName.equalsIgnoreCase(CONSOLE_APP_NAME) ||
+                applicationName.equalsIgnoreCase(MY_ACCOUNT_APP_NAME)) {
```
Comment (chamathns) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java:1045]:
```suggestion
        if (CONSOLE_APP_NAME.equalsIgnoreCase(applicationName) ||
                MY_ACCOUNT_APP_NAME.equalsIgnoreCase(applicationName)) {
```

**Review by chamathns — DISMISSED** [review]

**Review by chamathns — APPROVED** [review]

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java (hunk 1 above)
Comment (Thisara-Welmilla) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java]: Addressed with commit 11d1d85f3a26f7c35e0e7edbf11c777d02851daf

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java (hunk 2 above)
Comment (Thisara-Welmilla) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java]: Addressed with commit 11d1d85f3a26f7c35e0e7edbf11c777d02851daf

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java (hunk 4 above)
Comment (Thisara-Welmilla) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java]: Addressed with commit 11d1d85f3a26f7c35e0e7edbf11c777d02851daf

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java (hunk 6 above)
Comment (Thisara-Welmilla) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java]: Addressed with commit 11d1d85f3a26f7c35e0e7edbf11c777d02851daf

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java (hunk 3 above)
Comment (Thisara-Welmilla) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java:1066]: Addressed with commit 11d1d85f3a26f7c35e0e7edbf11c777d02851daf

**Pull request merged**